File: service/user-tracking/domain/tracker/location_command/location_command_service_impl.py
from aiokafka import AIOKafkaProducer
from uuid import UUID
from domain.tracker.location_command.data_type.location_request import LocationRequest

from domain.tracker.location_command.location_command_service import LocationCommandService
from domain.tracker.location_command.location_command_repository import LocationCommandRepository
from core.location import Location
from core.location_message import LocationMessage

class LocationCommandServiceImpl(LocationCommandService):
    location_repository: LocationCommandRepository
    message_producer: AIOKafkaProducer

    def __init__(self, location_repository: LocationCommandRepository, message_producer: AIOKafkaProducer):
        self.location_repository = location_repository
        self.message_producer = message_producer

    async def update_location(self, user_id: UUID, location: LocationRequest) -> None:
        location = Location(
            latitude=location.latitude,
            longitude=location.longitude,
            timestamp=location.timestamp,
            accuracy=location.accuracy
        )

        await self.location_repository.insert_location(location)

        location_message = LocationMessage(
            id=location.id,
            user_id=user_id,
            latitude=location.latitude,
            longitude=location.longitude,
            timestamp=location.timestamp,
            accuracy=location.accuracy
        )

File: service/user-tracking/src/main/java/project/tracknest/usertracking/domain/tracker/locationcommand/LocationCommandServiceImpl.java
package project.tracknest.usertracking.domain.tracker.locationcommand;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import project.tracknest.usertracking.core.datatype.LocationMessage;
import project.tracknest.usertracking.core.entity.Location;
import project.tracknest.usertracking.core.entity.User;
import project.tracknest.usertracking.proto.lib.UpdateUserLocationRequest;

import java.time.Instant;
import java.time.OffsetDateTime;
import java.time.ZoneOffset;
import java.util.Optional;
import java.util.UUID;

@Service
@RequiredArgsConstructor
@Slf4j
public class LocationCommandServiceImpl implements LocationCommandService {
    private final LocationCommandRepository locationRepository;
    private final LocationMessageProducer messageProducer;
    private final LocationCommandUserRepository userRepository;

    @Override
    @Transactional
    public void updateUserLocation(UUID userId, String username, UpdateUserLocationRequest request) {
        Optional<User> userOpt = userRepository.findById(userId);
        if (userOpt.isEmpty()) {
            log.warn("User with id {} not found. Cannot update location.", userId);
            return;
        }

        OffsetDateTime timestamp = OffsetDateTime.ofInstant(
                Instant.ofEpochMilli(
                        request.getTimestampMs()),
                ZoneOffset.UTC);

        User user = userOpt.get();
        user.setConnected(true);
        user.setLastActive(timestamp);
        userRepository.save(user);

        Location location = Location.builder()
                .latitude(request.getLatitudeDeg())
                .longitude(request.getLongitudeDeg())
                .accuracy(request.getAccuracyMeter())
                .velocity(request.getVelocityMps())
                .id(Location.LocationId
                        .builder()
                        .userId(userId)
                        .timestamp(timestamp)
                        .build())
                .build();

        locationRepository.save(location);

        LocationMessage message = LocationMessage.builder()
                .userId(userId)
                .username(username)
                .latitudeDeg(request.getLatitudeDeg())
                .longitudeDeg(request.getLongitudeDeg())
                .accuracyMeter(request.getAccuracyMeter())
                .velocityMps(request.getVelocityMps())
                .timestampMs(request.getTimestampMs())
                .build();

        messageProducer.produce(message);

        log.info("Received request to update location command");
    }
}
